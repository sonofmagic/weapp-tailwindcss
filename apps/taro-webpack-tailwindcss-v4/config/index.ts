import type { UserConfigExport } from '@tarojs/cli'
import path from 'node:path'
import process from 'node:process'
import { defineConfig } from '@tarojs/cli'
import TsconfigPathsPlugin from 'tsconfig-paths-webpack-plugin'
import { UnifiedWebpackPluginV5 } from 'weapp-tailwindcss/webpack'
import devConfig from './dev'
import prodConfig from './prod'

// https://taro-docs.jd.com/docs/next/config#defineconfig-辅助函数
export default defineConfig<'webpack5'>(async (merge, { command, mode }) => {
  const baseConfig: UserConfigExport<'webpack5'> = {
    projectName: 'taro-webpack-tailwindcss-v4',
    date: '2025-2-23',
    designWidth: 750,
    deviceRatio: {
      640: 2.34 / 2,
      750: 1,
      375: 2,
      828: 1.81 / 2,
    },
    sourceRoot: 'src',
    outputRoot: 'dist',
    defineConstants: {
    },
    copy: {
      patterns: [
      ],
      options: {
      },
    },
    framework: 'react',
    compiler: {
      type: 'webpack5',
      prebundle: {
        enable: false,
      },
    },
    cache: {
      enable: false, // Webpack 持久化缓存配置，建议开启。默认配置请参考：https://docs.taro.zone/docs/config-detail#cache
    },
    terser: {
      enable: true,
      config: {

      },
    },
    mini: {
      postcss: {
        pxtransform: {
          enable: true,
          config: {

          },
        },
        cssModules: {
          enable: false, // 默认为 false，如需使用 css modules 功能，则设为 true
          config: {
            namingPattern: 'module', // 转换模式，取值为 global/module
            generateScopedName: '[name]__[local]___[hash:base64:5]',
          },
        },
      },
      webpackChain(chain) {
        chain.resolve.plugin('tsconfig-paths').use(TsconfigPathsPlugin)
        chain.merge({
          plugin: {
            install: {
              plugin: UnifiedWebpackPluginV5,
              args: [
                {
                  rem2rpx: true,
                  // tailwindcssBasedir: path.resolve(__dirname, '../'),
                  tailwindcss: {
                    // 显示声明用的是 tailwindcss v4
                    version: 4,
                    v4: {
                      cssEntries: [
                        // app.css 的路径
                        path.resolve(__dirname, '../src/app.css'),
                      ],
                    },
                  },
                },
              ],
            },
          },
        })
      },
    },
    h5: {
      publicPath: '/',
      staticDirectory: 'static',
      output: {
        filename: 'js/[name].[hash:8].js',
        chunkFilename: 'js/[name].[chunkhash:8].js',
      },
      miniCssExtractPluginOption: {
        ignoreOrder: true,
        filename: 'css/[name].[hash].css',
        chunkFilename: 'css/[name].[chunkhash].css',
      },
      postcss: {
        autoprefixer: {
          enable: true,
          config: {},
        },
        cssModules: {
          enable: false, // 默认为 false，如需使用 css modules 功能，则设为 true
          config: {
            namingPattern: 'module', // 转换模式，取值为 global/module
            generateScopedName: '[name]__[local]___[hash:base64:5]',
          },
        },
      },
      webpackChain(chain) {
        chain.resolve.plugin('tsconfig-paths').use(TsconfigPathsPlugin)
      },
    },
    rn: {
      appName: 'taroDemo',
      postcss: {
        cssModules: {
          enable: false, // 默认为 false，如需使用 css modules 功能，则设为 true
        },
      },
    },
  }

  process.env.BROWSERSLIST_ENV = process.env.NODE_ENV

  if (process.env.NODE_ENV === 'development') {
    // 本地开发构建配置（不混淆压缩）
    return merge({}, baseConfig, devConfig)
  }
  // 生产构建配置（默认开启压缩混淆等）
  return merge({}, baseConfig, prodConfig)
})
