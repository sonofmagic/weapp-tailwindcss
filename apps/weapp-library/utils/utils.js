import Promise from './es6-promise'

/**
 * 函数节流 函数连续调用时，func 执行频率限定为 次 / wait
 * underscore.js 实现
 * @param {Function} func 传入函数
 * @param {number} wait 表示时间窗口的间隔
 * @param {object} options 如果想忽略开始边界上的调用，传入{leading: false}
 * @param {object} options.leading 是否在开始边界上调用，默认true
 * @param {object} options.trailing 是否在结尾边界上调用，默认true
 * @return {Function} 返回客户调用函数
 */
export function throttle(func, wait, options) {
  let context, args, result
  let timeout = null
  // 上次执行时间点
  let previous = 0
  if (!options) { options = {} }
  // 延迟执行函数
  const later = function () {
    // 若设定了开始边界不执行选项，上次执行时间始终为0
    previous = options.leading === false ? 0 : new Date()
    timeout = null
    result = func.apply(context, args)
    if (!timeout) { context = args = null }
  }
  return function () {
    const now = new Date()
    // 首次执行时，如果设定了开始边界不执行选项，将上次执行时间设定为当前时间。
    if (!previous && options.leading === false) { previous = now }
    // 延迟执行时间间隔
    const remaining = wait - (now - previous)
    context = this
    args = arguments
    // 延迟时间间隔remaining小于等于0，表示上次执行至此所间隔时间已经超过一个时间窗口
    // remaining大于时间窗口wait，表示客户端系统时间被调整过
    if (remaining <= 0 || remaining > wait) {
      clearTimeout(timeout)
      timeout = null
      previous = now
      result = func.apply(context, args)
      if (!timeout) { context = args = null }
    // 如果延迟执行不存在，且没有设定结尾边界不执行选项
    }
    else if (!timeout && options.trailing !== false) {
      timeout = setTimeout(later, remaining)
    }
    return result
  }
}

/**
 * 函数防抖 返回函数连续调用时，空闲时间必须大于或等于 wait，func 才会执行
 * https://www.npmjs.com/package/debounce-promise
 * @param {Function} func 传入函数
 * @param {number} wait 表示时间窗口的间隔
 * @param {object} options 其他参数
 * @param {boolean} options.leading 调用触发于开始边界而不是结束边界
 * @param {boolean} options.accumulate 保留现在接收到的所有参数并以数组的方式传递给传入函数
 * @return {Function} 返回客户调用函数
 */
export function debounce(fn, wait = 0, options = {}) {
  let lastCallAt
  let deferred
  let timer
  let pendingArgs = []
  return function debounced(...args) {
    const currentWait = getWait(wait)
    const currentTime = new Date().getTime()

    const isCold = !lastCallAt || (currentTime - lastCallAt) > currentWait

    lastCallAt = currentTime

    if (isCold && options.leading) {
      return options.accumulate
        ? Promise.resolve(fn.call(this, [args])).then(result => result[0])
        : Promise.resolve(fn.call(this, ...args))
    }

    if (deferred) {
      clearTimeout(timer)
    }
    else {
      deferred = defer()
    }

    pendingArgs.push(args)
    timer = setTimeout(flush.bind(this), currentWait)

    if (options.accumulate) {
      const argsIndex = pendingArgs.length - 1
      return deferred.promise.then(results => results[argsIndex])
    }

    return deferred.promise
  }

  function flush() {
    const thisDeferred = deferred
    clearTimeout(timer)

    Promise.resolve(
      options.accumulate
        ? fn.call(this, pendingArgs)
        : fn.apply(this, pendingArgs[pendingArgs.length - 1]),
    )
      .then(thisDeferred.resolve, thisDeferred.reject)

    pendingArgs = []
    deferred = null
  }
}

function getWait(wait) {
  return (typeof wait === 'function') ? wait() : wait
}

function defer() {
  const deferred = {}
  deferred.promise = new Promise((resolve, reject) => {
    deferred.resolve = resolve
    deferred.reject = reject
  })
  return deferred
}

/**
 * 将日期对象转为日期字符串: '2018-04-22'
 * @param glue {String} 连接符号
 */
export function formatDate(date, glue = '-') {
  const year = date.getFullYear()
  const month = date.getMonth() + 1
  const day = date.getDate()
  return [year, month, day].map((n) => {
    n = n.toString()
    return n[1] ? n : `0${n}`
  }).join(glue)
}

/**
 * 日期加减
 * @param interval {String} 要添加的时间间隔的单位
 *        { y: year, q: quarter, M: month, w: week,
            d: day, h: hour, m: minute, s: second}
 * @param number {Number} 要添加的时间间隔
 * @param date 时间对象
 * @return {Date} 新的时间对象
 * 使用：
 * var now = new Date();
 * var newDate = DateAdd( "d", 5, now);
 */
export function dateAdd(interval, number, date) {
  switch (interval) {
    case 'y': {
      date.setFullYear(date.getFullYear() + number)
      return date
    }
    case 'q': {
      date.setMonth(date.getMonth() + number * 3)
      return date
    }
    case 'M': {
      date.setMonth(date.getMonth() + number)
      return date
    }
    case 'w': {
      date.setDate(date.getDate() + number * 7)
      return date
    }
    case 'd': {
      date.setDate(date.getDate() + number)
      return date
    }
    case 'h': {
      date.setHours(date.getHours() + number)
      return date
    }
    case 'm': {
      date.setMinutes(date.getMinutes() + number)
      return date
    }
    case 's': {
      date.setSeconds(date.getSeconds() + number)
      return date
    }
  }
}
