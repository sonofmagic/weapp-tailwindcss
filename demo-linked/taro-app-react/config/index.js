// let TaroWeappTailwindcssWebpackPluginV5
let UnifiedWebpackPluginV5
const path = require('path')
const isLocal = process.env.LOCAL
const isWrite = process.env.WRITE
if (isLocal) {
  console.log('use local built webpack plugin')
  const { UnifiedWebpackPluginV5: plugin } = require('../weapp-tw-dist')
  UnifiedWebpackPluginV5 = plugin
} else {
  const { UnifiedWebpackPluginV5: plugin } = require('weapp-tailwindcss-webpack-plugin/webpack')
  UnifiedWebpackPluginV5 = plugin
}
const config = {
  compiler: {
    prebundle: {
      enable: false,
    },
    type: 'webpack5'
  },
  projectName: 'myApp',
  date: '2022-2-5',
  designWidth: 750,
  deviceRatio: {
    640: 2.34 / 2,
    750: 1,
    828: 1.81 / 2
  },
  cache: {
    enable: false
  },
  sourceRoot: 'src',
  outputRoot: 'dist',
  plugins: [],
  defineConstants: {},
  copy: {
    patterns: [],
    options: {}
  },
  framework: 'react',
  mini: {
    postcss: {
      pxtransform: {
        enable: true,
        config: {
          selectorBlackList: ['shadow']
        }
      },
      url: {
        enable: true,
        config: {
          limit: 1024 // 设定转换尺寸上限
        }
      },
      cssModules: {
        enable: false, // 默认为 false，如需使用 css modules 功能，则设为 true
        config: {
          namingPattern: 'module', // 转换模式，取值为 global/module
          generateScopedName: '[name]__[local]___[hash:base64:5]'
        }
      }
      // tailwindcss: {},
      // autoprefixer: {},
    },
    /**
     *
     * @param {import('webpack-chain')} chain
     * @param {import('webpack')} webpack
     */
    webpackChain(chain, webpack) {

      const opt = {
        appType: 'taro',
        minifiedJs: true
        // customAttributes: {
        //   // '*': ['emptyImageClass','btnClassName'],
        //   '*': [/Class/]
        // }
      }
      // chain.module.rules

      // chain.merge({
      //   plugin: {
      //     install: {
      //       plugin: TaroWeappTailwindcssWebpackPluginV5,
      //       args: [opt]
      //     }
      //   }
      // })
      if (process.env.TARO_ENV === 'weapp') {
        chain.merge({
          plugin: {
            install: {
              plugin: UnifiedWebpackPluginV5,
              // args: [opt, 'taro']
            }
          }
        })
      }


      // chain
      //   .plugin('UnifiedWebpackPlugin')
      //   .use(UnifiedWebpackPlugin, [opt])
      // chain
      //   .plugin('UnifiedWebpackPluginV5')
      //   .use(UnifiedWebpackPluginV5(opt))

      // if (isWrite) {
      //   opt.loaderOptions = {
      //     jsxRename: {
      //       dir: path.resolve(__dirname, '../../../test/fixtures/loader/taro-app')
      //     }
      //   }
      // }


    }
  },
  h5: {
    publicPath: '/',
    staticDirectory: 'static',
    postcss: {
      autoprefixer: {
        enable: true,
        config: {}
      },
      cssModules: {
        enable: false, // 默认为 false，如需使用 css modules 功能，则设为 true
        config: {
          namingPattern: 'module', // 转换模式，取值为 global/module
          generateScopedName: '[name]__[local]___[hash:base64:5]'
        }
      }
    }
  }
}

module.exports = function (merge) {
  if (process.env.NODE_ENV === 'development') {
    return merge({}, config, require('./dev'))
  }
  return merge({}, config, require('./prod'))
}
