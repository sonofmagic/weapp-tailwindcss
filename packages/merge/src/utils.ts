import type { ClassValue } from 'clsx'
import type {
  createTailwindMerge as _createTailwindMergeV3,
  extendTailwindMerge as _extendTailwindMergeV3,
  twJoin as _twJoinV3,
  twMerge as _twMergeV3,
} from 'tailwind-merge'
import type {
  createTailwindMerge as _createTailwindMergeV2,
  extendTailwindMerge as _extendTailwindMergeV2,
  twJoin as _twJoinV2,
  twMerge as _twMergeV2,
} from 'tailwind-merge-v2'
import type { CreateOptions } from './types'
import { clsx } from 'clsx'
import { escape } from 'weapp-tailwindcss/escape'

export type TailwindMergeVersion = 2 | 3
// 处理 escape 逻辑
export function noop(x: string) {
  return x
}

export interface CreateFactoryOptions<V extends TailwindMergeVersion> {
  version: V
  twMerge: V extends 3 ? typeof _twMergeV3 : typeof _twMergeV2
  twJoin: V extends 3 ? typeof _twJoinV3 : typeof _twJoinV2
  extendTailwindMerge: V extends 3 ? typeof _extendTailwindMergeV3 : typeof _extendTailwindMergeV2
  createTailwindMerge: V extends 3 ? typeof _createTailwindMergeV3 : typeof _createTailwindMergeV2
}

export function createFactory<V extends TailwindMergeVersion>(opts: CreateFactoryOptions<V>) {
  const {
    twMerge: _twMerge,
    twJoin: _twJoin,
    createTailwindMerge: _createTailwindMerge,
    extendTailwindMerge: _extendTailwindMerge,
  } = opts

  return function create(options?: CreateOptions) {
    const escapeFn = options?.escapeFn ?? escape
    const e = options?.disableEscape ? noop : escapeFn

    function twMerge(...inputs: ClassValue[]): string {
      return e(_twMerge(clsx(...inputs)))
    }

    function twJoin(...inputs: ClassValue[]): string {
      return e(_twJoin(clsx(...inputs)))
    }

    // 版本安全的 extendTailwindMerge
    const extendTailwindMerge = (...args: Parameters<typeof _extendTailwindMerge>) => {
      // @ts-ignore
      const customTwMerge = _extendTailwindMerge(...args)
      return function cn(...inputs: ClassValue[]): string {
        return e(customTwMerge(clsx(...inputs)))
      }
    }

    // 版本安全的 createTailwindMerge
    const createTailwindMerge = (...args: Parameters<typeof _createTailwindMerge>) => {
      // @ts-ignore
      const customTwMerge = _createTailwindMerge(...args)
      return function cn(...inputs: ClassValue[]): string {
        return e(customTwMerge(clsx(...inputs)))
      }
    }

    return {
      twMerge,
      twJoin,
      extendTailwindMerge,
      createTailwindMerge,
    }
  }
}
