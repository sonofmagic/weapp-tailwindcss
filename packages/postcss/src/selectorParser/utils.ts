// 选择器 AST 相关的通用工具，提供组合与替换的辅助函数
import type { Node } from 'postcss-selector-parser'
import type { IStyleHandlerOptions } from '../types'
import psp from 'postcss-selector-parser'

export type ParserTransformOptions = Partial<{
  lossless: boolean
  updateSelector: boolean
}>

// normalizeTransformOptions 确保 parser 在修改选择器时保持必要的副作用
export function normalizeTransformOptions(options?: ParserTransformOptions): ParserTransformOptions {
  return {
    lossless: false,
    updateSelector: true,
    ...options,
  }
}

// mklist 将节点扩展为 "节点 + combinator + 节点副本" 的结构
export function mklist(node: Node): Node[] {
  return [
    node,
    psp.combinator({
      value: '+',
    }),
    node.clone(),
  ]
}

// composeIsPseudoAst 按需把字符串数组转换为 :is(...) 结构的 AST
export function composeIsPseudoAst(strs: string | string[]): Node[] {
  if (typeof strs === 'string') {
    return mklist(psp.tag({
      value: strs,
    }))
  }
  if (strs.length > 1) {
    return mklist(psp.pseudo({
      value: ':is',
      nodes: strs.map(str =>
        psp.tag({
          value: str,
        }),
      ),
    }))
  }
  return mklist(psp.tag({
    value: strs[0],
  }))
}

// 根据配置生成替换子代选择器的 AST，默认使用 view 标签
export function getCombinatorSelectorAst(options: IStyleHandlerOptions) {
  let childCombinatorReplaceValue: Node[] = mklist(psp.tag({ value: 'view' }))
  const { cssChildCombinatorReplaceValue } = options
  if (
    typeof cssChildCombinatorReplaceValue === 'string'
    || (Array.isArray(cssChildCombinatorReplaceValue) && cssChildCombinatorReplaceValue.length > 0)) {
    childCombinatorReplaceValue = composeIsPseudoAst(cssChildCombinatorReplaceValue)
  }
  return childCombinatorReplaceValue
}
