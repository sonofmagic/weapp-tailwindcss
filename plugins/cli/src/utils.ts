import type { Transform } from 'node:stream'
import fs from 'fs-extra'

const cssMatcher = (file: string) => /.+\.(?:wx|ac|jx|tt|q|c)ss$/.test(file)
const htmlMatcher = (file: string) => /.+\.(?:(?:(?:wx|ax|jx|ks|tt|q)ml)|swan)$/.test(file)
const jsMatcher = (file: string) => {
  if (file.includes('node_modules')) {
    return false
  }
  return /.+\.[cm]?[jt]s?$/.test(file)
}

export { cssMatcher, htmlMatcher, jsMatcher }

export function touch(filename: string) {
  const time = new Date()

  try {
    fs.utimesSync(filename, time, time)
  } catch {
    fs.closeSync(fs.openSync(filename, 'w'))
  }
}

export { default as set } from 'set-value'
export { default as get } from 'get-value'

export function isObject(x: unknown): x is Record<string | symbol | number, unknown> {
  return typeof x === 'object' && x !== null
}

export function promisify(task: NodeJS.ReadWriteStream | NodeJS.ReadWriteStream[]): Promise<unknown> {
  return new Promise((resolve, reject) => {
    if (Array.isArray(task)) {
      return Promise.all(task.map((x) => promisify(x)))
        .then(resolve)
        .catch(reject)
    } else {
      if ((<Transform>task).destroyed) {
        resolve(undefined)
        return
      }
      task
        .on('finish', () => {
          resolve(undefined)
        })
        .on('error', (err) => {
          reject(err)
        })
    }
  })
}

export function arrify<T>(val: T) {
  return Array.isArray(val) ? (val as T) : [val]
}
