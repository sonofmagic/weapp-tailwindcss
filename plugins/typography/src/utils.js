const isPlainObject = require('lodash.isplainobject')

const parser = require('postcss-selector-parser')
const parseSelector = parser()

module.exports = {
  isUsableColor(color, values) {
    return isPlainObject(values) && color !== 'gray' && values[600]
  },

  /**
   * @param {string} selector
   */
  commonTrailingPseudos(selector) {
    const ast = parseSelector.astSync(selector)

    /** @type {import('postcss-selector-parser').Pseudo[][]} */
    const matrix = []

    // Put the pseudo elements in reverse order in a sparse, column-major 2D array
    for (const [i, sel] of ast.nodes.entries()) {
      for (const [j, child] of [...sel.nodes].reverse().entries()) {
        // We only care about pseudo elements
        if (child.type !== 'pseudo' || !child.value.startsWith('::')) {
          break
        }

        matrix[j] = matrix[j] || []
        matrix[j][i] = child
      }
    }

    const trailingPseudos = parser.selector()

    // At this point the pseudo elements are in a column-major 2D array
    // This means each row contains one "column" of pseudo elements from each selector
    // We can compare all the pseudo elements in a row to see if they are the same
    for (const pseudos of matrix) {
      // It's a sparse 2D array so there are going to be holes in the rows
      // We skip those
      if (!pseudos) {
        continue
      }

      const values = new Set(pseudos.map((p) => p.value))

      // The pseudo elements are not the same
      if (values.size > 1) {
        break
      }

      for (const pseudo of pseudos) pseudo.remove()
      trailingPseudos.prepend(pseudos[0])
    }

    if (trailingPseudos.nodes.length > 0) {
      return [trailingPseudos.toString(), ast.toString()]
    }

    return [null, selector]
  }
}
