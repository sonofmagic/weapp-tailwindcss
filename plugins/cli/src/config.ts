import path from 'node:path'
import { cosmiconfigSync } from 'cosmiconfig'
import defu from 'defu'
import type { UserDefinedOptions } from 'weapp-tailwindcss'
import fs from 'fs-extra'
import { get, set } from '@/utils'

export type WeappTwCosmiconfigResult = {
  config: UserConfig
  filepath: string
  isEmpty?: boolean
}

export function createConfigLoader(root: string) {
  const explorer = cosmiconfigSync('weapp-tw')

  function search(searchFrom: string = root): WeappTwCosmiconfigResult | undefined {
    const searchFor = explorer.search(searchFrom)
    if (searchFor) {
      searchFor.config = defu<UserConfig, UserConfig[]>(searchFor.config, getDefaultConfig(root))
      return searchFor
    }
  }

  function load(filepath: string) {
    return explorer.load(filepath)
  }

  return {
    search,
    load
  }
}

export function getDefaultConfig(root: string): UserConfig {
  return {
    outDir: 'dist',
    root,
    srcDir: '.'
  }
}

export type UserConfig = {
  outDir?: string
  root?: string
  srcDir?: string
  weappTailwindcssOptions?: UserDefinedOptions
  clean?: boolean
}

export function defineConfig(options: UserConfig) {
  return options
}

export interface InitConfigOptions {
  lang?: 'js' | 'ts'
  root?: string
}

export function updateProjectConfig(options: { root: string; dest?: string }) {
  const { root, dest } = options
  const projectConfigFilename = 'project.config.json'
  const projectConfigPath = path.resolve(root, projectConfigFilename)
  if (fs.existsSync(projectConfigPath)) {
    try {
      const projectConfig = fs.readJSONSync(projectConfigPath) as {
        miniprogramRoot?: string
        srcMiniprogramRoot?: string
        setting: {
          packNpmManually: boolean
          packNpmRelationList: {
            packageJsonPath: string
            miniprogramNpmDistDir: string
          }[]
        }
      }

      set(projectConfig, 'miniprogramRoot', 'dist/')
      set(projectConfig, 'srcMiniprogramRoot', 'dist/')
      set(projectConfig, 'setting.packNpmManually', true)

      if (Array.isArray(get(projectConfig, 'setting.packNpmRelationList'))) {
        const x = projectConfig.setting.packNpmRelationList.find((x) => x.packageJsonPath === './package.json' && x.miniprogramNpmDistDir === './dist')
        if (!x) {
          projectConfig.setting.packNpmRelationList.push({
            packageJsonPath: './package.json',
            miniprogramNpmDistDir: './dist'
          })
        }
      } else {
        set(projectConfig, 'setting.packNpmRelationList', [
          {
            packageJsonPath: './package.json',
            miniprogramNpmDistDir: './dist'
          }
        ])
      }
      fs.outputJSONSync(dest ?? projectConfigPath, projectConfig, {
        spaces: 2
      })

      console.log(`✨ 设置 ${projectConfigFilename} 配置文件成功!`)
    } catch {
      console.warn(`✨ 设置 ${projectConfigFilename} 配置文件失败!`)
    }
  } else {
    console.warn(`✨ 没有找到 ${projectConfigFilename} 文件!`)
  }
}

export function updatePackageJson(options: { root: string; dest?: string }) {
  const { root, dest } = options
  const packageJsonFilename = 'package.json'
  const packageJsonPath = path.resolve(root, packageJsonFilename)
  if (fs.existsSync(packageJsonPath)) {
    try {
      const packageJson = fs.readJSONSync(packageJsonPath) as {
        scripts: Record<string, string>
      }
      set(packageJson, 'scripts.dev', 'weapp-tw dev')
      set(packageJson, 'scripts.build', 'weapp-tw build')
      fs.outputJSONSync(dest ?? packageJsonPath, packageJson, {
        spaces: 2
      })
    } catch {}
  }
}

export function initConfig(options?: InitConfigOptions) {
  const { lang, root } = defu<InitConfigOptions, InitConfigOptions[]>(options, { lang: 'js', root: process.cwd() }) as Required<InitConfigOptions>
  const configFilename = `weapp-tw.config.${lang ?? 'js'}`
  const configPath = path.resolve(root, configFilename)

  // const tsconfigPath = path.resolve(root, 'tsconfig.json')

  fs.ensureDirSync(root)
  if (lang === 'ts') {
    fs.writeFileSync(
      configPath,
      `import { defineConfig } from '@weapp-tailwindcss/cli'

export default defineConfig({})
`,
      'utf8'
    )
  } else {
    fs.writeFileSync(
      configPath,
      `/** @type {import('@weapp-tailwindcss/cli').UserConfig} */
module.exports = {}
`,
      'utf8'
    )
  }
  console.log(`✨ ${configFilename} 配置文件，初始化成功!`)

  updateProjectConfig({ root })
  updatePackageJson({ root })
  return configPath
}
