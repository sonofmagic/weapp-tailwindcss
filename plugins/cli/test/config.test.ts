import path from 'node:path'
import fs from 'fs-extra'
import { fixturesPath } from './utils'
import { createConfigLoader, initConfig, updateProjectConfig, updatePackageJson } from '@/config'

describe('config', () => {
  it('load js config', () => {
    const root = path.resolve(fixturesPath, 'configs/js')
    const configLoader = createConfigLoader(root)
    const s = configLoader.search()
    delete s?.config.root
    expect(s?.config).toMatchSnapshot()
  })

  it('load ts config', () => {
    const root = path.resolve(fixturesPath, 'configs/ts')
    const configLoader = createConfigLoader(root)
    const s = configLoader.search()
    delete s?.config.root
    expect(s?.config).toMatchSnapshot()
  })

  it('load nothing', () => {
    const root = path.resolve(fixturesPath, 'configs/nothing')
    const configLoader = createConfigLoader(root)
    const s = configLoader.search()
    delete s?.config.root
    expect(s?.config).toMatchSnapshot()
  })

  describe('initConfig', () => {
    it('init js config', () => {
      const root = path.resolve(fixturesPath, 'configs-output/js')
      const p = initConfig({
        root
      })
      expect(fs.existsSync(p)).toBe(true)
      const configLoader = createConfigLoader(root)
      const s = configLoader.search()
      delete s?.config.root
      expect(s?.config).toMatchSnapshot()
    })

    it('init ts config', () => {
      const root = path.resolve(fixturesPath, 'configs-output/ts')
      const p = initConfig({
        root,
        lang: 'ts'
      })
      expect(fs.existsSync(p)).toBe(true)
      const configLoader = createConfigLoader(root)
      const s = configLoader.search()
      delete s?.config.root
      expect(s?.config).toMatchSnapshot()
    })

    it('init js config with tsconfig', () => {
      const root = path.resolve(fixturesPath, 'configs-output/js-tsconfig')
      const p = initConfig({
        root
      })
      expect(fs.existsSync(p)).toBe(true)
      const configLoader = createConfigLoader(root)
      const s = configLoader.search()
      delete s?.config.root
      expect(s?.config).toMatchSnapshot()
    })

    it('init ts config with tsconfig', () => {
      const root = path.resolve(fixturesPath, 'configs-output/ts-tsconfig')
      const p = initConfig({
        root,
        lang: 'ts'
      })
      expect(fs.existsSync(p)).toBe(true)
      const configLoader = createConfigLoader(root)
      const s = configLoader.search()
      delete s?.config.root
      expect(s?.config).toMatchSnapshot()
    })
  })

  describe('updateProjectConfig', () => {
    it('updateProjectConfig js native project', async () => {
      const root = path.resolve(fixturesPath, 'configs/json/js')
      const target = path.resolve(root, 'output.json')
      updateProjectConfig({
        root,
        dest: target
      })
      expect(fs.existsSync(target)).toBe(true)
      const res = await import(target)
      expect(res.default).toMatchSnapshot()
    })

    it('updateProjectConfig js native configed project', async () => {
      const root = path.resolve(fixturesPath, 'configs/json/js-configed')
      const target = path.resolve(root, 'output.json')
      updateProjectConfig({
        root,
        dest: target
      })
      expect(fs.existsSync(target)).toBe(true)
      const res = await import(target)
      expect(res.default).toMatchSnapshot()
    })
  })

  describe('updatePackageJson', () => {
    it('updatePackageJson js native project', async () => {
      const root = path.resolve(fixturesPath, 'configs/js')
      const target = path.resolve(root, 'pkg.json')
      updatePackageJson({
        root,
        dest: target
      })
      expect(fs.existsSync(target)).toBe(true)
      const res = await import(target)
      expect(res.default).toMatchSnapshot()
    })

    it('updatePackageJson ts native project', async () => {
      const root = path.resolve(fixturesPath, 'configs/ts')
      const target = path.resolve(root, 'pkg.json')
      updatePackageJson({
        root,
        dest: target
      })
      expect(fs.existsSync(target)).toBe(true)
      const res = await import(target)
      expect(res.default).toMatchSnapshot()
    })
  })
})
