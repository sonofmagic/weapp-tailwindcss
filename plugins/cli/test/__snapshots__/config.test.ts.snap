// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`config > initConfig > init js config 1`] = `
{
  "outDir": "dist",
  "src": ".",
}
`;

exports[`config > initConfig > init js config with tsconfig 1`] = `
{
  "outDir": "dist",
  "src": "./miniprogram",
}
`;

exports[`config > initConfig > init ts config 1`] = `
{
  "outDir": "dist",
  "src": ".",
}
`;

exports[`config > initConfig > init ts config with tsconfig 1`] = `
{
  "outDir": "dist",
  "src": "./miniprogram",
}
`;

exports[`config > load js config 1`] = `
{
  "outDir": "dist",
  "src": ".",
}
`;

exports[`config > load nothing 1`] = `undefined`;

exports[`config > load ts config 1`] = `
{
  "outDir": "dist",
  "src": ".",
}
`;

exports[`config > updatePackageJson > updatePackageJson js native project 1`] = `
{
  "author": "",
  "dependencies": {
    "@weapp-tailwindcss/cli": "link:../../../..",
  },
  "description": "",
  "keywords": [],
  "license": "ISC",
  "main": "index.js",
  "name": "js",
  "scripts": {
    "build": "weapp-tw build",
    "dev": "weapp-tw dev",
    "postinstall": "weapp-tw patch",
    "test": "echo "Error: no test specified" && exit 1",
  },
  "version": "1.0.0",
}
`;

exports[`config > updatePackageJson > updatePackageJson ts native project 1`] = `
{
  "author": "",
  "dependencies": {
    "@weapp-tailwindcss/cli": "link:../../../..",
  },
  "description": "",
  "keywords": [],
  "license": "ISC",
  "main": "index.js",
  "name": "ts",
  "scripts": {
    "build": "weapp-tw build",
    "dev": "weapp-tw dev",
    "postinstall": "weapp-tw patch",
    "test": "echo "Error: no test specified" && exit 1",
  },
  "version": "1.0.0",
}
`;

exports[`config > updateProjectConfig > updateProjectConfig js native configed project 1`] = `
{
  "appid": "wxb3d842a4a7e3440d",
  "compileType": "miniprogram",
  "condition": {},
  "editorSetting": {
    "tabIndent": "auto",
    "tabSize": 2,
  },
  "libVersion": "trial",
  "miniprogramRoot": "dist/",
  "packOptions": {
    "ignore": [],
    "include": [],
  },
  "setting": {
    "babelSetting": {
      "disablePlugins": [],
      "ignore": [],
      "outputPath": "",
    },
    "coverView": true,
    "enhance": true,
    "es6": true,
    "minified": true,
    "packNpmManually": true,
    "packNpmRelationList": [
      {
        "miniprogramNpmDistDir": "./dist",
        "packageJsonPath": "./package.json",
      },
    ],
    "postcss": true,
    "showShadowRootInWxmlPanel": true,
  },
  "srcMiniprogramRoot": "dist/",
}
`;

exports[`config > updateProjectConfig > updateProjectConfig js native project 1`] = `
{
  "appid": "wxb3d842a4a7e3440d",
  "compileType": "miniprogram",
  "condition": {},
  "editorSetting": {
    "tabIndent": "auto",
    "tabSize": 2,
  },
  "libVersion": "trial",
  "miniprogramRoot": "dist/",
  "packOptions": {
    "ignore": [],
    "include": [],
  },
  "setting": {
    "babelSetting": {
      "disablePlugins": [],
      "ignore": [],
      "outputPath": "",
    },
    "coverView": true,
    "enhance": true,
    "es6": true,
    "minified": true,
    "packNpmManually": true,
    "packNpmRelationList": [
      {
        "miniprogramNpmDistDir": "./dist",
        "packageJsonPath": "./package.json",
      },
    ],
    "postcss": true,
    "showShadowRootInWxmlPanel": true,
  },
  "srcMiniprogramRoot": "dist/",
}
`;
